rids/rids.py

```diff
@@ -15,113 +15,111 @@
-        # TODO check format of ioc source, branch to different parsers
-        # For now, the only source format is newline-seprated bad IPv4 addresses
-        # TODO perform parsing in a separate function from config handling
-        for line in ioc_data.readlines():
-          line = line.strip().decode('utf-8')
-          if not line: continue
-          bad_ip = str(ipaddress.ip_address(line))
-          if bad_ip not in bad_ips:
-            bad_ips[bad_ip] = [source['name']]
-          else:
-            bad_ips[bad_ip].append(source['name'])
-  return bad_ips
-
-
-def get_external_ip():
-  with urllib.request.urlopen('https://ipinfo.io/ip') as my_ip:
-    return my_ip.read().decode("utf-8").strip()
+flags.DEFINE_string('config_path', '/etc/rids/config.json',
+                    'file path indicating where IOC configuration can be found')
+flags.DEFINE_string('eventlog_path', '/var/rids/events.log',
+                    'file path indicating where to append new event logs')
 
 
 def main(argv):
+  """Main entry point of the app.  Also bound to CLI `rids` by setuptools."""
   if len(argv) > 1:
     raise app.UsageError('Too many command-line arguments.')
 
+  logging.basicConfig(level=logging.DEBUG,
+                      format='%(asctime)s %(levelname)-8s %(message)s',
+                      datefmt='%m-%d %H:%M',
+                      filename=FLAGS.eventlog_path,
+                      filemode='a')
+
+  cfg = config.ParseFile(FLAGS.config_path)
+  ruleset = rules.RuleSet()
+  for ioc_config in cfg['iocs']:
+    ruleset.MergeRuleset(iocs.parse_ruleset(ioc_config))
+
+  host_ip = _get_host_ip()
+  print(f'Using {host_ip} as host IP address')
+
+  event_queue = queue.Queue()
+  threading.Thread(
+      target=_inspect_tls_traffic,
+      args=(host_ip, ruleset, event_queue))
+  threading.Thread(
+      target=_inspect_remote_endpoints,
+      args=(host_ip, ruleset, event_queue))
+
+  while True:
+    # We just log the suspicious events for now, but here is where we could
+    # do post-processing and/or share sightings of known / unknown threats.
+    event = event_queue.get()
+    logging.log(json.puts(event))
+    event_queue.task_done()
+
+
+def _inspect_tls_traffic(host_ip, ruleset, q):
+  """Worker function for thread that inspects client and server hellos.
+  
+  Args:
+    ruleset: a RuleSet object to perform evaluation with
+    q: a Queue for relaying events back to the main thread
+  """
+  tls_capture = network_capture.HandshakeScanner(host_ip)
+  for observation in tls_capture.scan():
+    events = ruleset.ProcessHandshake(observation)
+    for event in events:
+      q.put(event)
+
+
+def _inspect_remote_endpoints(host_ip, ruleset, q):
+  """Worker function for thread that inspects remote IP addresses.
+
+  Args:
+    ruleset: a RuleSet object to perform evaluation with
+    q: a Queue for relaying events back to the main thread
+  """
+  remote_ip_capture = network_capture.RemoteServerSanner(host_ip)
```

Comment:
We need linting/type checker to catch these. Too easy to have bugs like this.
```suggestion
  remote_ip_capture = network_capture.RemoteServerScanner(host_ip)
```